cashu/mint/ledger.py

```diff
@@ -209,35 +212,89 @@ def _verify_proof_bdhke(self, proof: Proof):
+                now = time.time()
+                if timelock < now:
+                    logger.trace(f"p2sh timelock ran out ({timelock}<{now}).")
+                    return True
+                logger.trace(f"p2sh timelock still active ({timelock}>{now}).")
+            except Exception as e:
+                # we skip on error for secrets without timelocks
+                logger.error(f"p2sh timelock invalid: {e}")
+                logger.trace(proof.secret)
+
+            # execute and verify P2SH
+            txin_p2sh_address, valid = verify_bitcoin_script(
+                proof.script.script, proof.script.signature
+            )
+            if not valid:
+                raise Exception("script invalid.")
             # check if secret commits to script address
             # format: P2SH:<address>:<secret>
-            assert len(proof.secret.split(":")) == 3, "secret format wrong."
+            assert len(proof.secret.split(":")) >= 3, "p2sh secret format invalid."
             assert proof.secret.split(":")[1] == str(
                 txin_p2sh_address
             ), f"secret does not contain correct P2SH address: {proof.secret.split(':')[1]} is not {txin_p2sh_address}."
-        return valid
+            return True
+
+        # P2PK
+        # proof.secret format: P2PK:<pubkey_hex>:<timelock>:<op_return>:<nonce>
+        # first we check whether the timelock is in the past
+        if proof.secret.startswith("P2PK:"):
+            try:
+                timelock = int(proof.secret.split(":")[2])
+                # check if timelock is in the past
+                now = time.time()
+                if timelock < now:
+                    logger.trace(f"p2pk timelock ran out ({timelock}<{now}).")
+                    return True
+                logger.trace(f"p2pk timelock still active ({timelock}>{now}).")
+            except Exception as e:
+                # we skip on error for secrets without timelocks
+                logger.error(f"p2pk timelock invalid: {e}")
+                logger.trace(proof.secret)
+
+            # now we check the signature
+            if not proof.p2pksig:
+                # no signature present although secret indicates one
+                raise Exception("no p2pk signature in proof.")
+
+            # we parse the secret as a P2PK commitment
+            assert len(proof.secret.split(":")) == 5, "p2pk secret format invalid."
+            pubkey = proof.secret.split(":")[1]
+
+            # check signature proof.p2pksig against pubkey
+            # we expect the signature to be on the pubkey (=message) itself
+            assert verify_p2pk_signature(
+                message=proof.secret.encode("utf-8"),
+                pubkey=PublicKey(bytes.fromhex(pubkey), raw=True),
+                signature=bytes.fromhex(proof.p2pksig),
+            ), "p2pk signature invalid."
+            logger.trace(proof.p2pksig)
+            logger.trace("p2pk signature valid.")

```

Comment:
Mint: P2PK verification. If timelock is in the past, the script condition is considered valid. If not, the signature on every token is checked.